### PR TITLE
policy: gpu: validation of vfio passthrough GPUs

### PR DESCRIPTION
**Pull request by manuelh-dev**
This work stream enables adding security policy to GPU CI tests by introducing support for VFIO GPU passthrough in genpolicy and rules.rego for both hot- and cold-plug scenarios. Description of changes:
- change pod.rs/policy.rs to: read NVIDIA GPU resource limits, translating these to the expected VFIO devices and expected OCI CDI annotations kata-agent will observe.
- change rules.rego to:
  - differentiate between volume devices and vfio devices
  - validate VFIO devices
  - validate VFIO CDI annotations (esp., make sure annotations and device entries match)
- add cold- and hot-plug tests to genpolicy testing utility
- add policy enforcement to NVIDIA GPU samples

Example for pass-through of a single NVIDIA GPU:
```
...
      limits:
        nvidia.com/pgpu: "1"
...
```
Resulting in about the following device and annotation entries observed by the agent in the hot-plug case (CreateContainerRequest):
- device entry with: `"container_path": "/dev/vfio/<num>"`, for example  `"container_path": "/dev/vfio/50"`
- annotation: `cdi.k8s.io/vfio<num>: nvidia.com/gpu=<num>`, for example `cdi.k8s.io/vfio50: nvidia.com/gpu=0`

For the cold-plug case, no VFIO device entry is expected (CreateContainerRequest), but CDI annotations are expected.

The implementation *should* be fairly straightforward to add other VFIO device types. The implementation is restrictive. For example, if a malicious device plugin attempts to cause injection of unexpected devices (e.g., `/dev/iommu`) or environment variables, the policy will reject it.

Parts of this PR will still need to change due to the use of a new device plugin referenced [here](https://github.com/kata-containers/kata-containers/pull/12087), but this version should be sufficiently good for a first round of feedback. Thanks to @danmihai1 already for having made a brief, preliminary pass.

### CODE REVIEW

File: src/tools/genpolicy/rules.rego [hunk 1]
```diff
@@ -461,12 +479,47 @@ allow_devices(p_devices, i_devices) if {
...
+    print("allow_device 1: start")
+    some p_device in p_devices
+    p_device.container_path == i_device.container_path
+    print("allow_device 1: true")
+}
+
+# TODO: can this go into genpolicy and will this persist with the new device plugin approach?
+# Allow VFIO PCI guest kernel devices for GPU passthrough (dynamically added)
+# Pattern: /dev/vfio/<num> where <num> is one or more digits
+allow_device(p_devices, i_device) if {
```
Comment (danmihai1) [inline src/tools/genpolicy/rules.rego]: This looks reasonable at a quick look - unless the "vfio-pci-gk" deserves to be read from genpolicy-settings.json. Do we expect other device type strings too, in the future? That would make it a good candidate for genpolicy-settings.json.

File: src/tools/genpolicy/src/policy.rs [hunk 2]
```diff
@@ -1092,6 +1107,18 @@ fn get_container_annotations(
...
             .or_insert(image_name);
+
+        // Add CDI annotation regex pattern for GPU containers
+        // The actual device numbers are assigned at runtime by the shim
+        // Pattern matches: cdi.k8s.io/vfio<num> = nvidia.com/gpu=<num>
+        if let Some(nvidia_pgpu_count) = yaml_container.get_nvidia_pgpu_count() {
+            if nvidia_pgpu_count > 0 {
+                annotations.insert(
+                    "^cdi\\.k8s\\.io/vfio[0-9]+$".to_string(),
+                    "^nvidia\\.com/gpu=[0-9]+$".to_string(),
```
Comment (danmihai1) [inline src/tools/genpolicy/src/policy.rs]: "^nvidia\\.com/gpu=[0-9]+$" should be read by genpolicy from genpolicy-settings.json - that way hopefully supporting other vendor regex strings in the future too. An example of handling regex strings through genpolicy-settings.json:

            "forbidden_source_regex": [
                "^(?:0{0,4}:){0,7}0{0,3}1$",
                "^127\\.(?:[0-9]{1,3}\\.){2}[0-9]{1,3}$"
            ]

File: src/tools/genpolicy/rules.rego [hunk 3]
```diff
@@ -259,6 +263,20 @@ allow_anno_key(i_key, p_oci) if {
 
     print("allow_anno_key 2: true")
 }
+allow_anno_key(i_key, p_oci) if {
+    print("allow_anno_key 3 (regex): i key =", i_key)
+
+    some p_key_regex, p_value_regex in p_oci.Annotations
```
Comment (danmihai1) [inline src/tools/genpolicy/rules.rego]: I think I understand why you used the Annotations field here, but to me it seems cleaner to store those policy regular expressions in the Policy data outside the OCI data structure.

How about a Policy data field similar to:

      "exec_commands": []

And construct that field value based on:
- Just the input YAML if that makes sense.
- If needed a genpolicy-settings.json field similar to:

    "ExecProcessRequest": {
      "regex": []
    },

File: src/tools/genpolicy/rules.rego [hunk 4]
```diff
@@ -457,16 +475,107 @@ allow_log_directory(p_oci, i_oci) if {
...
+    count(vfio_numbers) == count({n | n := vfio_numbers[_]})
+
+    print("allow_vfio_device_count: true")
+}
+
+
+allow_vfio_cdi_correlation(i_vfio_devices, i_oci) if {
+    print("allow_vfio_cdi_correlation: start")
+
+    vfio_numbers := [suffix |
```
Comment (danmihai1) [inline src/tools/genpolicy/rules.rego:560]: Is this copy/paste from the function above? If yes, maybe these two functions should be merged to a single function?

File: src/tools/genpolicy/rules.rego (hunk 1 above)
Comment (zvonkok) [reply src/tools/genpolicy/rules.rego]: Make sure we support both `vfio_mode=guest-kernel` and `vfio_mode=vfio`

File: src/tools/genpolicy/tests/policy/testdata/createcontainer/gpu_vfio_cdi/testcases.json [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+          "io.katacontainers.pkg.oci.bundle_path": "/run/containerd/io.containerd.runtime.v2.task/k8s.io/gpu-container",
+          "io.katacontainers.pkg.oci.container_type": "pod_container",
+          "io.kubernetes.cri.container-name": "gpu-container",
+          "io.kubernetes.cri.container-type": "container",
+          "io.kubernetes.cri.image-name": "registry.k8s.io/pause@sha256:3d380ca8864549e74af4b29c10f9cb0956236dfb01c40ca076fb6c37253234db",
+          "io.kubernetes.cri.sandbox-id": "abcd1234567890abcdef1234567890abcdef1234567890abcdef1234567890ab",
+          "io.kubernetes.cri.sandbox-name": "gpu-test-pod",
+          "io.kubernetes.cri.sandbox-namespace": "default",
+          "io.kubernetes.cri.sandbox-uid": "12345678-1234-1234-1234-123456789012"
+        },
```
[the remark this answers is not in the document]
Comment (zvonkok) [reply src/tools/genpolicy/tests/policy/testdata/createcontainer/gpu_vfio_cdi/testcases.json:654]: Indeed, the `cdi.k8s.io/vfio<NUM>` NUM is derived either from the IOMMU group or IOMMUFD NUM, you cannot have the same `vfio1` for different devices

**Review by lifupan — CHANGES_REQUESTED** [review]
Hi @manuelh-dev  @zvonkok 

Alex has basically finished implementing policy support for runtime-rs, and currently all CI tests for the policy are passing on runtime-rs. Therefore, we must ensure that any policy modifications you make are also supported on runtime-rs and that all CI tests pass; otherwise, runtime-rs will never be able to guarantee that all non-TEE CI tests will pass.

File: tests/integration/kubernetes/k8s-nvidia-nim.bats [hunk 6]
```diff
@@ -79,6 +79,8 @@ setup_kbs_credentials() {
 create_inference_pod() {
     envsubst <"${POD_INSTRUCT_YAML_IN}" >"${POD_INSTRUCT_YAML}"
 
+    auto_generate_policy "${policy_settings_dir}" "${pod_instruct_yaml}"
```
[the remark this answers is not in the document]
Comment (manuelh-dev) [reply tests/integration/kubernetes/k8s-nvidia-nim.bats]: Appeared after rebase, acknowledging

File: tests/integration/kubernetes/k8s-nvidia-nim.bats [hunk 7]
```diff
@@ -94,6 +96,8 @@ create_inference_pod() {
 create_embedqa_pod() {
     envsubst <"${POD_EMBEDQA_YAML_IN}" >"${POD_EMBEDQA_YAML}"
 
+    auto_generate_policy "${policy_settings_dir}" "${pod_embedqa_yaml}"
```
[the remark this answers is not in the document]
Comment (manuelh-dev) [reply tests/integration/kubernetes/k8s-nvidia-nim.bats]: Appeared after rebase, acknowledging

File: src/tools/genpolicy/tests/policy/testdata/createcontainer/gpu_vfio_cdi/testcases.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (manuelh-dev) [reply src/tools/genpolicy/tests/policy/testdata/createcontainer/gpu_vfio_cdi/testcases.json:654]: Yup, this is an AI-generated "negative" policy integration test case for the policy validation functions. Looks like this scenario where bad/overlapping keys are passed cannot be reflected with this framework (due to the JSON structure). I will rework/remove this test.

File: src/tools/genpolicy/rules.rego (hunk 3 above)
Comment (manuelh-dev) [reply src/tools/genpolicy/rules.rego]: @danmihai1 - for clarification:
In the policy generation phase, we add a single p_oci annotation key-value regex pair when a GPU is detected. This helps this function to iterate over all i_oci annotations key-value pairs and regex-match them.
This function here is currently authored to be oblivious to the NVIDIA GPU use case, to generally add support for regex key-value pairs.

I see two options now with your suggestion:
1) make the function CDI-annotation specific, less "general-purpose". We could try to detect if specific GPU devices are in i_devices, and then match annotations which so far didn't pass checks against our known GPU key-value regex pair specific to the device type being plugged. This way we wouldn't need to add any field (not into p_oci.Annotations and not into an extra field we would create).
2) keep the function "general-purpose" and use a field, as you suggested, in which we can generally place annotation regexes which we want to allow for input annotations. The reasoning for these to be in an extra field would then be that these are annotations which cannot be clearly predicted, hence not being added to p_oci.Annotations, but being added into the extra field.

If you have preferences, please let me know. :)

File: src/tools/genpolicy/genpolicy-settings.json [hunk 8]
```diff
@@ -197,6 +197,14 @@
             ]
         }
     },
+    "device_annotations": {
```
Comment (manuelh-dev) [inline src/tools/genpolicy/genpolicy-settings.json:200]: @danmihai1 - moved various strings here while keeping others at the top of rules.rego. Suggest to assess whether this choice makes sense to you as well. Other than that, I have addressed and resolved all other comments. The change set now looks quite different, but I think we're converging towards the right direction. Thanks a lot for all your help so far, and no rush.

File: src/tools/genpolicy/src/policy.rs [hunk 9]
```diff
@@ -682,6 +707,28 @@ impl AgentPolicy {
...
+        // Generate expected device entries and annotation key-value pairs for VFIO devices
+        let mut runtime_anno_patterns = BTreeMap::new();
+        if let Some(nvidia_pgpu_count) = yaml_container.get_nvidia_pgpu_count() {
+            if nvidia_pgpu_count > 0 {
+                for _ in 0..nvidia_pgpu_count {
+                    let mut device = agent::Device::new();
+                    // TODO: change device_path to /dev/vfio/devices/vfio<num> when the new device plugin is used.
+                    // The actual device number /dev/vfio/<num> is assigned at runtime by the device plugin
+                    device.set_container_path(self.config.settings.device_annotations.vfio.device_path.clone());
+                    device.set_type(self.config.settings.device_annotations.vfio.nvidia_gpu_gk_device_type.clone());
```
Comment (manuelh-dev) [inline src/tools/genpolicy/src/policy.rs]: @zvonkok - RE a prior comment of yours: at policy generate time, is there a chance to differentiate between whether we expect `vfio_mode=guest-kernel` and when to expect `vfio_mode=vfio`? If not, I believe we need to make this an array. And either allow `vfio-pci-gk` or `vfio-pci`

File: tests/integration/kubernetes/tests_common.sh [hunk 10]
```diff
@@ -300,6 +300,28 @@ add_requests_to_policy_settings() {
 	done
 }
 
+# Temporary function: Add support for CDI environment variables
+# used for GPU device passthrough (with current device plugin).
+#
+# Pattern: PCI_RESOURCE_<VENDOR>_<DEVICE>=<PCI_ADDRESS>
+# PCI address format: 4 hex digits : 2 hex digits : 2 hex digits . 1 digit
+add_cdi_envvars_to_policy_settings() {
```
Comment (manuelh-dev) [inline tests/integration/kubernetes/tests_common.sh]: @zvonkok - at what point will this environment variable no longer be added? Should this already be the case with the recent cold-plug PR, or are further changes required to get rid of this variable?

File: src/tools/genpolicy/genpolicy-settings.json (hunk 8 above)
Comment (danmihai1) [reply src/tools/genpolicy/genpolicy-settings.json:200]: If we're adding here

"device_path": "/dev/vfio",

then we should use that as a replacement for 

VFIO_DEVICE_PATH = "/dev/vfio"

I can't think of good reasons to keep both of these definitions.

File: src/tools/genpolicy/rules.rego [hunk 11]
```diff
@@ -259,6 +264,17 @@ allow_anno_key(i_key, p_oci) if {
 
     print("allow_anno_key 2: true")
 }
+allow_anno_key(i_key, p_oci) if {
+    print("allow_anno_key 3: i key =", i_key)
+
+    some p_key_regex, p_value_regex in policy_data.containers[_].runtime_anno_patterns
```
Comment (danmihai1) [inline src/tools/genpolicy/rules.rego]: This seems wrong:

`some p_key_regex, p_value_regex in policy_data.containers[_].runtime_anno_patterns`

because you are searching through the policy containers again. But, the caller is already operating on a particular policy container - so there is no need to search again. If I understand correctly, we need something similar to this. Please make these changes carefully because I am not sure my quick syntax here is correct:

```
allow_anno(p_container, i_oci)
...
allow_anno(p_container, i_oci) if {
    print("allow_anno 1: start")

    not i_oci.Annotations

    print("allow_anno 1: true")
}
allow_anno(p_container, i_oci) if {
    p_oci := p_container.OCI

    print("allow_anno 2: p Annotations =", p_oci.Annotations)
    print("allow_anno 2: i Annotations =", i_oci.Annotations)

    every i_key, i_value in i_oci.Annotations {
        allow_anno_key_value(i_key, i_value, p_container)
    }

    print("allow_anno 2: true")
}

allow_anno_key_value(i_key, i_value, p_container) if {
    print("allow_anno_key 1: i key =", i_key)

    startswith(i_key, "io.kubernetes.cri.")

    print("allow_anno_key 1: true")
}
allow_anno_key_value(i_key, i_value, p_container) if {
    print("allow_anno_key 2: i key =", i_key)

    some p_key, _ in p_container.OCI.Annotations
    p_key == i_key

    print("allow_anno_key 2: true")
}
allow_anno_key_value(i_key, i_value, p_container) if {
    print("allow_anno_key 3: i key =", i_key, "i_value =", i_value)

    some p_key_regex, p_value_regex in p_container.runtime_anno_patterns

    print("allow_anno_key 3: p_key_regex =", p_key_regex, "p_value_regex =", p_value_regex)
    regex.match(p_key_regex, i_key)
    regex.match(p_value_regex, i_value)

    print("allow_anno_key 3: true")
}
```

File: src/tools/genpolicy/genpolicy-settings.json (hunk 8 above)
Comment (manuelh-dev) [reply src/tools/genpolicy/genpolicy-settings.json:200]: Good point - I have also removed the notion of VFIO_DEVICE_PREFIX

File: src/tools/genpolicy/rules.rego (hunk 11 above)
Comment (manuelh-dev) [reply src/tools/genpolicy/rules.rego]: You're absolutely right, thank you, addressed

File: src/tools/genpolicy/src/policy.rs (hunk 9 above)
Comment (zvonkok) [reply src/tools/genpolicy/src/policy.rs]: For now only allow vfio-pci-gk we do not have any plans to support DPDK/SPDK like workloads.

File: src/tools/genpolicy/rules.rego [hunk 12]
```diff
@@ -223,41 +225,50 @@ concat_op_if_not_null(ops, op) = result if {
...
 
-    print("allow_anno_key 2: true")
+    print("allow_anno_key_value 2: true")
+}
+allow_anno_key_value(i_key, i_value, p_container) if {
+    print("allow_anno_key_value 3: i key =", i_key, "i_value =", i_value)
+
+    some p_key_regex, p_value_regex in p_container.runtime_anno_patterns
+
+    regex.match(p_key_regex, i_key)
```
Comment (danmihai1) [inline src/tools/genpolicy/rules.rego:269]: I would add after the "some" line:

print("allow_anno_key_value 3: p_key_regex = ", p_key_regex, "p_value_regex = " p_value_regex)

File: src/tools/genpolicy/rules.rego [hunk 13]
```diff
@@ -457,16 +468,93 @@ allow_log_directory(p_oci, i_oci) if {
...
-        print("allow_devices: i_device =", i_device)
-        some p_device in p_devices
-        p_device.container_path == i_device.container_path
-    }
+
+    vfio_device_path := policy_data.device_annotations.vfio.device_path
+
+    p_volume_devices := [d | d := p_devices[_]; d.container_path != vfio_device_path]
+    i_volume_devices := [d | d := i_devices[_]; not startswith(d.container_path, vfio_device_path)]
+    allow_volume_devices(p_volume_devices, i_volume_devices)
```
Comment (danmihai1) [inline src/tools/genpolicy/rules.rego:480]: Print the two collections after calculating them - here and for vfio devices below.

File: src/tools/genpolicy/rules.rego [hunk 14]
```diff
@@ -457,16 +468,93 @@ allow_log_directory(p_oci, i_oci) if {
...
+
+    count(i_vfio_devices) == count(p_vfio_devices)
+
+    vfio_device_path := policy_data.device_annotations.vfio.device_path
+    vfio_numbers := [suffix |
+        d := i_vfio_devices[_];
+        suffix := trim_prefix(d.container_path, vfio_device_path);
+        regex.match("^[0-9]+$", suffix)
+    ]
+    count(vfio_numbers) == count({n | n := vfio_numbers[_]})
```
Comment (danmihai1) [inline src/tools/genpolicy/rules.rego:566]: Is this count expression always true?

File: src/tools/genpolicy/rules.rego [hunk 15]
```diff
@@ -457,16 +468,93 @@ allow_log_directory(p_oci, i_oci) if {
...
+
+    cdi_suffixes := [suffix |
+        some key, _ in i_oci.Annotations;
+        startswith(key, CDI_VFIO_ANNOTATION_PREFIX);
+        suffix := trim_prefix(key, CDI_VFIO_ANNOTATION_PREFIX);
+        regex.match("^[0-9]+$", suffix)
+    ]
+
+    count(vfio_numbers) == count(cdi_suffixes)
+    {n | n := vfio_numbers[_]} == {s | s := cdi_suffixes[_]}
```
Comment (danmihai1) [inline src/tools/genpolicy/rules.rego:570]: Can the count expression above be removed? The next line seems to cover verifying the count already.

File: tests/integration/kubernetes/tests_common.sh (hunk 10 above)
Comment (manuelh-dev) [reply tests/integration/kubernetes/tests_common.sh]: Added documentation header.

File: src/tools/genpolicy/rules.rego (hunk 14 above)
Comment (manuelh-dev) [reply src/tools/genpolicy/rules.rego:566]: The idea of this check was to prevent the OCI runtime from sending duplicate /dev/vfio<X> entries, while the idea of the prior/first count expression was to assure the number of input and policy device entries match. That, so far, makes sense to me. :)

File: src/tools/genpolicy/rules.rego (hunk 15 above)
Comment (manuelh-dev) [reply src/tools/genpolicy/rules.rego:570]: You're right, I removed that line, however I added the line:
`count(cdi_suffixes) == count({s | s := cdi_suffixes[_]})`

This way, similar to vfio device number suffixes we validate that the CDI number suffixes are unique to e.g. prevent this from passing: cdi_suffixes = ["0", "1", "1"]

So, both vfio device number suffixes and CDI number suffixes need to be unique, and, the sets need to match.

File: src/tools/genpolicy/rules.rego (hunk 14 above)
Comment (danmihai1) [reply src/tools/genpolicy/rules.rego:566]: @manuelh-dev , my understanding is that vfio_numbers cannot have duplicate entries, because in Rego "Sets are unordered collections of **unique** values." Please add a test case in genpolicy's "make test" if you still think we need this count() == count() test.

File: src/tools/genpolicy/rules.rego (hunk 15 above)
Comment (danmihai1) [reply src/tools/genpolicy/rules.rego:570]: Duplicate entries in cdi_suffixes would seem to me like a bug in Regorus. Please add a test case for that in case you think Regorus might allow such duplicates.

File: src/tools/genpolicy/rules.rego (hunk 14 above)
Comment (manuelh-dev) [reply src/tools/genpolicy/rules.rego:566]: Added a test. We concluded that the left hand side uses an array (with duplicates) and the right-hand side uses the duplicate-filtered set.

File: src/tools/genpolicy/rules.rego (hunk 15 above)
Comment (manuelh-dev) [reply src/tools/genpolicy/rules.rego:570]: As discussed in another review comment, I have added a test case for now.

Leaving this comment open with regards to our conclusion on whether we aim to validate against duplicate entries being passed or not.

Comment (danmihai1) [reply src/tools/genpolicy/rules.rego:570]: @manuelh-dev explained to me the difference between sets & arrays, so now these rules seem correct. to me too. 

Manuel, it might be helpful to add a comment here - e.g., "convert array to set, to reject possible duplicate entries in the array".

**Comment by manuelh-dev** [comment]
GPU-SNP test has following pre-requisite: https://github.com/kata-containers/kata-containers/pull/12175

**Comment by manuelh-dev** [comment]
@lifupan - to my untrained eye, the runtime-rs test failures in the latest CI run seem unrelated to these changes here - please let me know if you think otherwise.

File: src/tools/genpolicy/tests/policy/testdata/createcontainer/gpu_vfio_cdi/testcases.json [hunk 16]
```diff
@@ -0,0 +1,847 @@
...
+          "vm_path": ""
+        },
+        {
+          "container_path": "/dev/vfio/2",
+          "id": "2",
+          "options": ["0000:66:00.0=06/00"],
+          "type_": "vfio-pci-gk",
+          "vm_path": ""
+        }
+      ],
```
Comment (fidencio) [inline src/tools/genpolicy/tests/policy/testdata/createcontainer/gpu_vfio_cdi/testcases.json:97]: This code is not taking into consideration IOMMUFD, where we'd receive something like `/dev/vfio/devices/vfio0`.

Can we make sure it'll work for that case as well?

Comment (manuelh-dev) [reply src/tools/genpolicy/tests/policy/testdata/createcontainer/gpu_vfio_cdi/testcases.json:97]: Per discussion with @zvonkok , it will be an either/or. So, once the shift to `/dev/vfio/devices/vfio0` paths is complete, I will especially update `src/tools/genpolicy/genpolicy-settings.json`:
`"device_path": "/dev/vfio/",` to the newly expected path. With this, I will also need to update the tests

**Comment by manuelh-dev** [comment]
The GPU-SNP CI task which makes use of this new policy capability is still failing. We are awaiting a change on our CI runner to fix this. The PR is ready though. Adding `do-not-merge` meanwhile.

File: src/tools/genpolicy/rules.rego [hunk 17]
```diff
@@ -457,16 +469,123 @@ allow_log_directory(p_oci, i_oci) if {
...
+allow_vfio_device(p_vfio_devices, i_vfio_device) if {
+    print("allow_vfio_device: start")
+
+    some p_device in p_vfio_devices
+
+    vfio_device_path := policy_data.device_annotations.vfio.device_path
+    startswith(i_vfio_device.container_path, vfio_device_path)
+    suffix := trim_prefix(i_vfio_device.container_path, vfio_device_path)
+    regex.match("^[0-9]+$", suffix)
+
```
[the remark this answers is not in the document]
Comment (manuelh-dev) [reply src/tools/genpolicy/rules.rego:522]: ID [0, INT_MAX] ...

File: src/tools/genpolicy/src/policy.rs [hunk 18]
```diff
@@ -682,6 +723,54 @@ impl AgentPolicy {
...
+                        self.config
+                            .settings
+                            .device_annotations
+                            .vfio
+                            .nvidia_gpu_gk_device_type
+                            .clone(),
+                    );
+                    device.set_vm_path("".to_string());
+                    devices.push(device);
+                }
```
[the remark this answers is not in the document]
Comment (manuelh-dev) [reply src/tools/genpolicy/src/policy.rs:756]: Revised the comment above. Thank you.

File: src/tools/genpolicy/src/policy.rs [hunk 19]
```diff
@@ -672,6 +701,18 @@ impl AgentPolicy {
         let mut devices: Vec<agent::Device> = vec![];
         if let Some(volumeDevices) = &yaml_container.volumeDevices {
             for volumeDevice in volumeDevices {
+                if volumeDevice
+                    .devicePath
+                    .starts_with(&self.config.settings.device_annotations.vfio.device_path)
+                {
+                    panic!(
+                        "Volume device path '{}' conflicts with VFIO device path prefix '{}'. \
+                         VFIO devices should be requested via resource limits (e.g., nvidia.com/gpu), not as volume devices.",
```
[the remark this answers is not in the document]
Comment (manuelh-dev) [reply src/tools/genpolicy/src/policy.rs]: Revised the comment, thank you

**Comment by fidencio** [comment]
I'm removing @lifupan's blocker as the tests pass on runtime-rs, meaning that we're not introducing any regression there.
